web/ui/react-app/src/pages/graph/PanelList.tsx

```diff
@@ -1,219 +1,178 @@
-        }, []),
-      },
-      this.updateURL
-    );
+    ]);
   };
 
-  render() {
-    const { metricNames, pastQueries, timeDriftError, fetchMetricsError, panels } = this.state;
-    const { pathPrefix } = this.props;
-    return (
-      <>
-        <Row className="mb-2">
-          <Checkbox
-            id="query-history-checkbox"
-            wrapperStyles={{ margin: '0 0 0 15px', alignSelf: 'center' }}
-            onChange={this.toggleQueryHistory}
-            defaultChecked={this.isHistoryEnabled()}
-          >
-            Enable query history
-          </Checkbox>
-          <Checkbox
-            id="use-local-time-checkbox"
-            wrapperStyles={{ margin: '0 0 0 15px', alignSelf: 'center' }}
-            onChange={this.toggleUseLocalTime}
-            defaultChecked={this.useLocalTime()}
-          >
-            Use local time
-          </Checkbox>
-        </Row>
-        <Row>
-          <Col>
-            {timeDriftError && (
-              <Alert color="danger">
-                <strong>Warning:</strong> Error fetching server time: {timeDriftError}
-              </Alert>
-            )}
-          </Col>
-        </Row>
-        <Row>
-          <Col>
-            {fetchMetricsError && (
-              <Alert color="danger">
-                <strong>Warning:</strong> Error fetching metrics list: {fetchMetricsError}
-              </Alert>
-            )}
-          </Col>
-        </Row>
-        {panels.map(({ id, options }) => (
-          <Panel
-            onExecuteQuery={this.handleExecuteQuery}
-            key={id}
-            options={options}
-            onOptionsChanged={opts => this.handleOptionsChanged(id, opts)}
-            useLocalTime={this.state.useLocalTime}
-            removePanel={() => this.removePanel(id)}
-            metricNames={metricNames}
-            pastQueries={pastQueries}
-            pathPrefix={pathPrefix}
-          />
-        ))}
-        <Button color="primary" className="add-panel-btn" onClick={this.addPanel}>
-          Add Panel
-        </Button>
-      </>
-    );
-  }
-}
+  return (
+    <>
+      {panels.map(({ id, options }) => (
+        <Panel
+          onExecuteQuery={handleExecuteQuery}
+          key={id}
+          options={options}
+          onOptionsChanged={opts =>
+            callAll(setPanels, updateURL)(panels.map(p => (id === p.id ? { ...p, options: opts } : p)))
+          }
+          removePanel={() =>
+            callAll(setPanels, updateURL)(
+              panels.reduce<PanelMeta[]>(
+                (acc, panel) => (panel.id !== id ? [...acc, { ...panel, key: `${acc.length}` }] : acc),
+                []
+              )
+            )
+          }
+          useLocalTime={useLocalTime}
+          metricNames={metrics}
+          pastQueries={queryHistoryEnabled ? historyItems : []}
+          pathPrefix={pathPrefix}
+        />
+      ))}
+      <Button color="primary" onClick={addPanel}>
```

Comment:
You lost the `className="add-panel-btn"` here, which is required to add the bottom margin for the button (when there's a vertical scrollbar, it looks better if you can scroll a bit further than the button, otherwise it always touches the bottom of the window).